### ISSUE TITLE
Creating new Engine takes very long

### ISSUE THREAD

**Issue by hstaudacher** [ui4j/ui4j#45]
I wanted to have isolated session using the BrowserEngine. What I did is to create a new `WebKitBrowserProvider`, creating engines, shutting down and create new ones. This is because I fount no other way to create isolated sessions. The second time the engine will be create takes ages. This is a unit test to reproduce. 

```java
@Test
  public void testCanCreateNewSession() {
    WebKitBrowserProvider provider = new WebKitBrowserProvider();
    provider.setShutdownListener( engine -> {} );
    BrowserEngine engine = provider.create();
    engine.navigate( "http://google.com" );
    engine.shutdown();
    BrowserEngine engine2 = provider.create();
    engine2.navigate( "http://google.com" );
  }
```
Anything you can do about it?

**Comment by hstaudacher** [comment]
Oh, I think it's the `shutdown` call that takes so long

**Comment by ui4j** [comment]
`WebKitBrowser.launchedJFX` flag is used to check if JavaFX ui thread is up and running. Its my mistake that after shutdown i forget to toggle the value. [This](https://github.com/ui4j/ui4j/commit/eb4877a9afcdec04e4364931b42520e94f98259e) commit fixed the bug. But its not enough to solve your problem. This time your sample application throws *IllegalStateException: Application launch must not be called more than once* error. I am going to find a way to run Application more than once.

**Comment by hstaudacher** [comment]
phew, ok. Do you think it's possible? Any other idea how to realize isolated sessions?

**Comment by hstaudacher** [comment]
Maybe to explain this usecase you can look at how ghostdriver has realized it: https://github.com/detro/ghostdriver/blob/master/test/java/src/test/java/ghostdriver/IsolatedSessionTest.java

**Comment by ui4j** [comment]
Similar problem exist for the TestFX project. I am going try to use the suggested [solution](https://github.com/TestFX/TestFX/issues/56).

**Comment by ui4j** [comment]
I understood the requirement after reading IsolatedSessionTest. Its not required to create multiple JavaFX application to solve this problem. Isolation means that new session must have an empty CookieStore. I add new method `clearCookies` to `BrowserEngine`. It just clear the all cookies from CookieStore to initialize new sessions. You could check the similar unit [test](https://github.com/ui4j/ui4j/blob/master/ui4j-webkit/src/test/java/com/ui4j/test/IsolatedSessionTest.java).

**Comment by hstaudacher** [comment]
I disagree. The clearCookies method is useless I think. With your Interception mechanism you have full control I guess.

Isolated Sessions means having multiple separated instances. Think about a multi user environment e.g. a WebApplication. A WebApplication that uses ui4j needs to have it's own (or multiple) sessions per user. Clearing and restoring cookies here is not an option. Especially because it prevents parallelism because you have to queue the usage of BrowserEngine.

**Comment by hstaudacher** [comment]
I'm not very familiar with JavaFX, but would it make sense to have one Application per JRE and use a new Window for every BrowserEngine instance? Is this possible at all? This would mean a shutdown does not eliminate the Application, instead it would eliminate the window.

**Comment by hstaudacher** [comment]
Any thoughts?

**Comment by ui4j** [comment]
Possible solutions:

> 1) Using custom CookieManager

Unfortunately CookiManager is static and shared per JVM so all connections share the same CookieStore

> 2) Using [custom](http://stackoverflow.com/questions/16305486/cookiemanager-for-multiple-threads) CookieStore per Thread

JavaFX WebView create a new thread for each connection. Each thread use its own CookieStore instead of shared CookieStore.

> 3) Using Ui4j interceptor mechanism

I try to manage cookies manually but somehow WebView still shared the cookies. I am trying to understand whats wrong about my implementation.

**Comment by hstaudacher** [comment]
Hmm, solution 2 sounds ok, isn't it?

**Comment by hstaudacher** [comment]
Hmm, after thinking about it the concept in UI4J that comes next to a session is the `Page`. Each page has it's own interceptor and handlers. So, the solution here would be to isolate Pages, right? This would remove the need to shutdown the engine and so on. So, every user could have the same engine but work on a different page that is isolated form each other. What do you think?

**Comment by hstaudacher** [comment]
Or there is a thing missing between Engine and Page e.g. a Session object.

**Comment by ui4j** [comment]
> So, the solution here would be to isolate Pages, right? 

Exactly

> This would remove the need to shutdown the engine and so on

Yes, the current implementation of the `BrowserEngine.shutdown` dispose JavaFX resources (threads etc) it has nothing to do with WebKit session. So its not necessary to call this method to clear the sessions.

> So, every user could have the same engine but work on a different page that is isolated form each other.

Yes our aim is isolating the Pages from each other but Java Cookie api doesnt let us to do this. `CookieHandler.setDefault()` is static and shared per JVM.

I am still trying to apply solution three to isolate the sessions.

**Comment by ui4j** [comment]
WebView open **5** connection per page. This could be configured with **http.maxConnections** system property. If you set the value **1** solution 2 works as expected for [the unit test](https://github.com/ui4j/ui4j/blob/master/ui4j-webkit/src/test/java/com/ui4j/test/IsolatedSessionTest.java). This might be used for simple unit test scenarios but not suitable for sites thats execute lots of ajax calls.

**Comment by ui4j** [comment]
Unfortunately solution 2 fails for real life usage, i should find a better way to solve this problem.

**Comment by hstaudacher** [comment]
hmm, what about this:

1) Create a custom `UI4JCookieManager extends CookieManager`
2) Hold one instance per `BrowserEngine` (maybe there is only one per application)
3) Everytime requests are made ensure it's set as the default cookie manager
4) Save the old cookie manager within the `UI4JCookieManager` and use it as a delegate
5) Within the put/get methods find a way to store the cookie by page (this is probably the hard part). The key in the `CookieStore` is an URL. So we need to enhance it to also know to which Page it belongs. 
6) Delegate all other cookie put/get requests to the default manager. 

So, basically this is just intercepting the default `CookieManager`. Don't know if this can work. What do you think?

**Comment by hstaudacher** [comment]
Btw. I'm not the first one with this idea ;)

http://stackoverflow.com/questions/15583057/using-a-hacked-cookiemanager-to-allow-2-different-instances-of-webview-with-diff

**Comment by hstaudacher** [comment]
Btw, why does solution 2 fail? Have you tried the ThreadLocal approach?

**Comment by ui4j** [comment]
`com.ui4j.webkit.WebKitIsolatedCookieHandler` could be used for the isolated sessions. This new CookieHandler works both gmail and JIRA. The implementation is independent from ui4j and could be used without dependency. This cookie handler always must be set before the JavaFX initialization.

**Comment by hstaudacher** [comment]
Ok sounds good. After looking at the Handler I noticed that a memory leak is pre programmed. The handle only provides a `clear` method that empties the two static maps. But if a page is closed the webview and it's handler stay in the map. This is a leak :)

**Comment by ui4j** [comment]
The latest commit prevents memory leaks. `Page.close` method removes `WebView` from static maps.
